File: ts/architecture/src/entity.ts
/** Entities and aggregate roots. */

import type { Identifier } from './identifier';

/**
 * A thing with a stable identity over time; equality is by id, not by field
 * values. In Rust this is an object-unsafe trait; in TS it is a plain interface,
 * and "equality by id" is `a.id === b.id` (branded ids compare by value).
 */
export interface Entity<TId extends Identifier = Identifier> {
  readonly id: TId;
}

/**
 * The transactional consistency boundary — the only kind of type a
 * {@link Repository} loads or stores.
 *
 * Rust models the human-facing name as a compile-time `const NAME` reachable
 * without an instance. TS interfaces cannot require statics ergonomically, so it
 * becomes a `readonly name` instance member. If you need the name without an
 * instance (e.g. for a `NotFound` error), expose it as a `static` on your class
 * as well.
 */
export interface AggregateRoot<TId extends Identifier = Identifier> extends Entity<TId> {
  /** Stable, human-facing name used for `NotFound { entity, .. }` errors and logs. */
  readonly name: string;
}

File: src/architecture/entity.rs
//! Entities and aggregate roots.

use super::identifier::Identifier;

/// A thing with a stable identity over time; equality is by id, not by field
/// values. Object-*unsafe* (associated `Id`, `Self`-shaped contract): entities
/// are concrete, so this is implemented directly, never used behind `dyn`.
///
/// # Examples
///
/// ```
/// use ev::architecture::{Entity, Id};
///
/// struct OrderTag;
/// type OrderId = Id<OrderTag, u64>;
///
/// struct Order {
///     id: OrderId,
///     total_cents: u64,
/// }
///
/// impl Entity for Order {
///     type Id = OrderId;
///     fn id(&self) -> OrderId {
///         self.id
///     }
/// }
///
/// let order = Order { id: OrderId::from_raw(1), total_cents: 999 };
/// assert_eq!(order.id(), OrderId::from_raw(1));
/// ```
pub trait Entity {
	type Id: Identifier;

	fn id(&self) -> Self::Id;
}

/// The transactional consistency boundary — the only kind of type a
/// [`super::repository::Repository`] loads or stores. A marker: it adds intent,
/// not methods. It deliberately does *not* require an event type (see
/// [`super::event::EmitsEvents`]), so a non-event aggregate is truly zero-cost.
///
/// # Examples
///
/// ```
/// use ev::architecture::{AggregateRoot, Entity, Id};
///
/// struct OrderTag;
/// type OrderId = Id<OrderTag, u64>;
///
/// struct Order {
///     id: OrderId,
/// }
///
/// impl Entity for Order {
///     type Id = OrderId;
///     fn id(&self) -> OrderId {
///         self.id
///     }
/// }
///
/// impl AggregateRoot for Order {
///     const NAME: &'static str = "order";
/// }
///
/// // `NAME` is a compile-time constant, reachable without an instance — e.g. for
/// // a `NotFound { entity: Order::NAME, .. }` error.
/// assert_eq!(Order::NAME, "order");
/// ```
pub trait AggregateRoot: Entity {
	/// Stable, human-facing name used for `NotFound { entity, .. }` errors and
	/// logs.
	const NAME: &'static str;
}
